cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -154,32 +161,66 @@ public void onRemoval(RemovalNotification<ApplicationId, Injector> notification)
             if (runner instanceof Service) {
               stopQuietly((Service) runner);
             }
+            removePreviewDir(injector.getInstance(Key.get(ProgramId.class,
+                                                          Names.named(PreviewRunnerModule.PREVIEW_PROGRAM_ID))));
           }
-          ApplicationId application = notification.getKey();
-          if (application == null) {
-            return;
-          }
-          removePreviewDir(application);
         }
       })
       .build();
   }
 
+  @Override
+  protected void startUp() throws Exception {
+    File previewDir = previewDataDir.toFile();
+
+    // there should be at most 10 directories so the process should not take very long
+    for (File file : DirUtils.listFiles(previewDir)) {
+      if (!file.isDirectory()) {
+        continue;
+      }
+      ProgramId programId;
+      try {
+        programId = ProgramId.fromString(PROGRAM_PREFIX + file.getName());
+      } catch (Exception e) {
+        // if there is an exception converting to a preview id, just continue
```

Comment:
Better log the exception. If the exception is really unimportant, at least log a debug message with the file name.